### PR TITLE
Install napari from repository in docker image

### LINKED ISSUE #6096: [ghcr.io/napari/napari-xpra] napari:0.4.18 contains napari 0.4.17
## 🐛 Bug

The docker container labeled 0.4.18 actually has napari 0.4.17 inside installed by pip.
cc: @jaimergp 

## To Reproduce

1. `docker run -it --rm -p 9876:9876 ghcr.io/napari/napari-xpra:0.4.18`
2. use the Terminal of the container: `pip show napari`

## Expected behavior

A container labeled 0.4.18 should have matching napari version.

## Environment

N/A

## Additional context

As discussed in the packaging meeting on July 26, 2023, this is probably due to time lag between napari getting to pypi and building the container.

### PR DESCRIPTION
**Pull request by jaimergp**
# Fixes/Closes

Closes https://github.com/napari/napari/issues/6096

# Description

We have a workflow that runs on each push and publishes a docker image to ghcr.io. This includes tagged releases.

However, the Dockerfile installs napari from PyPI, which will always be lagging. Hence, the docker container for 0.4.18 ships 0.4.17... and so do all the previous commits since 0.4.17 excluded 😂  We have been publishing the same Docker image for each push for a while and maaaybe we should delete a few (although layers are cached, so not so bad).

# References

- `build-args`: https://github.com/docker/build-push-action/tree/v4/#inputs
- ARG vs ENV: https://vsupalov.com/docker-arg-env-variable-guide/#arg-and-env-availability

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Maintenance (changes required to run napari, tests, & CI smoothly)
- [ ] Enhancement (non-breaking improvements of an existing feature)
- [ ] Documentation (update of docstrings and deprecation information)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

- [x] CI passes and reports correct version  

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by jaimergp** [comment]
commit 95f04d2 contains the CI for Docker, with [successful version tests](https://github.com/napari/napari/actions/runs/5690285943/job/15423343141?pr=6097#step:6:21).

File: dockerfile [hunk 1]
```diff
@@ -6,6 +6,7 @@ FROM --platform=linux/amd64 ubuntu:22.04 AS napari
 # below env var required to install libglib2.0-0 non-interactively
 ENV TZ=America/Los_Angeles
 ARG DEBIAN_FRONTEND=noninteractive
+ARG NAPARI_COMMIT=main
```
Comment (Czaki) [inline dockerfile:9]: This provides only default value?

Comment (jaimergp) [reply dockerfile:9]: Yep, if not provided via CLI, this (`main`) is the default value.

File: dockerfile [hunk 2]
```diff
@@ -33,8 +34,9 @@ RUN apt-get update && \
         libxcb-shape0 \
         && apt-get clean
 
-# install napari release version
-RUN pip3 install napari[all]
+# install napari from repo
+# see https://github.com/pypa/pip/issues/6548#issuecomment-498615461 for syntax
+RUN pip3 install "napari[all] @ git+https://github.com/napari/napari.git@${NAPARI_COMMIT}"
```
Comment (jaimergp) [inline dockerfile:39]: Maybe we can also install from the already checked out repo. We just need to pass it in the context (if not already passed). In other words:

```suggestion
RUN pip3 install ".[all]"
```

Comment (Czaki) [reply dockerfile:39]: it will not work until we copy the repo content to the docker container.

Comment (jaimergp) [reply dockerfile:39]: Then let's leave it like this. I think it's simpler.

**Review by Czaki — APPROVED** [review]

**Review by Czaki — APPROVED** [review]

**Pull request merged**